File: LC/java/Copy List with Random Pointer.java
/**
 * Definition for singly-linked list with a random pointer.
 * class RandomListNode {
 *     int label;
 *     RandomListNode next, random;
 *     RandomListNode(int x) { this.label = x; }
 * };
 */
public class Solution {
    public RandomListNode copyRandomList(RandomListNode head) {
        RandomListNode dummy,tmp,p;
        HashMap<RandomListNode,RandomListNode> map = new HashMap<RandomListNode,RandomListNode>();
        dummy = new RandomListNode(0);
        tmp =dummy;
        
        if (head == null){
            return null;
        }
        while (head!=null){
            if(map.containsKey(head)){
                p = map.get(head);
            }
            else{
                p = new RandomListNode(head.label);
                map.put(head,p);
            }
            tmp.next = p;
            if(head.random!=null){
                if(map.containsKey(head.random)){
                    p.random = map.get(head.random);
                }else{
                    p.random = new RandomListNode(head.random.label);
                    map.put(head.random,p.random);
                }
            }
            tmp = p;
            head = head.next;
        }
        return dummy.next;
    }
}

File: LC/python/Copy List with Random Pointer.py
# Definition for singly-linked list with a random pointer.
# class RandomListNode:
#     def __init__(self, x):
#         self.label = x
#         self.next = None
#         self.random = None

class Solution:
    # @param head, a RandomListNode
    # @return a RandomListNode
    def copyRandomList(self, head):
        if head== None:
            return None
        tmp = head
        
        while(tmp):
            p = RandomListNode(tmp.label)
            p.next = tmp.next
            tmp.next =p
            tmp = tmp.next.next
            
        tmp = head
        while tmp:
            if tmp.random:
                tmp.next.random = tmp.random.next
            tmp = tmp.next.next
        
        newhead = head.next
        old = head
        new = newhead
        while new.next:
            old.next = new.next
            old = old. next
            new.next = old.next
            new = new.next
        old.next = None
        new.next = None
        return newhead
            
